[Clang][AArch64] Implement widening FMMLA intrinsics

Amichaxx opened the pull request:
Proposed in [this ACLE proposal](https://github.com/ARM-software/acle/pull/409), this PR implements widening FMMLA intrinsics.

- F16 to F32
- MF8 to F32
- MF8 to F16

Additional changes:
- IsOverloadCvt flag renamed to IsOverloadFirstandLast for clarity, as the name implies conversion. Implementation remains unchanged.

Amichaxx committed:
Implement widening FMMLA intrinsics

- F16 to F32
- MF8 to F32
- MF8 to F16

Lukacma commented on clang/include/clang/Basic/arm_sve.td:
```diff
@@ -1196,6 +1196,18 @@ def SVMLLA_F32 : SInst<"svmmla[_f32]", "dddd","f", MergeNone, "aarch64_sve_fmmla
+}
+
+let SVETargetGuard = "sve2,f8f32mm", SMETargetGuard = InvalidMode in {
+  def SVMLLA_F32_MF8 : SInst<"svmmla[_f32_mf8]", "MM~~>", "m", MergeNone, "aarch64_sve_fmmla_mf8f32", [IsOverloadNone]>;
```
```suggestion
  def SVMLLA_F32_MF8 : SInst<"svmmla[_f32_mf8]", "dd~~>", "f", MergeNone, "aarch64_sve_fmmla_mf8f32", [IsOverloadNone]>;
```

Lukacma commented on clang/include/clang/Basic/arm_sve.td:
```diff
@@ -1196,6 +1196,18 @@ def SVMLLA_F32 : SInst<"svmmla[_f32]", "dddd","f", MergeNone, "aarch64_sve_fmmla
+}
+
+let SVETargetGuard = "sve2,f8f16mm", SMETargetGuard = InvalidMode in {
+  def SVMLLA_F16_MF8 : SInst<"svmmla[_f16_mf8]", "OO~~>", "m", MergeNone, "aarch64_sve_fmmla_mf8f16", [IsOverloadNone]>;
```
```suggestion
  def SVMLLA_F16_MF8 : SInst<"svmmla[_f16_mf8]", "dd~~>", "h", MergeNone, "aarch64_sve_fmmla_mf8f16", [IsOverloadNone]>;
```

Amichaxx committed:
- Implemented overloading for fmmla intrinsics, replaced fixed-type intrinsics
- Prototype cleanups
- Updated ll tests to remove unnecessary IR
- Removed unused arguments in clang test macros
- Removed redundant check lines in ll tests

Amichaxx commented on llvm/test/CodeGen/AArch64/sve-fmmla-f32f16.ll:
```diff
@@ -0,0 +1,32 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 6
+; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sve-f16f32mm < %s | FileCheck %s --check-prefixes=CHECK
```
Done, thanks.

Amichaxx committed:
Added new fp8 intrinsic in IntrinsicsAArch64.td, updated tests accordingly
Formatting

Amichaxx commented on llvm/include/llvm/IR/IntrinsicsAArch64.td:
```diff
@@ -2805,7 +2805,11 @@ def int_aarch64_sve_sudot_lane : AdvSIMD_SVE_DOT_Indexed_Intrinsic;
+def int_aarch64_sve_fmmla
+  : DefaultAttrsIntrinsic<[ llvm_anyvector_ty ],
+                          [ LLVMMatchType<0>, llvm_anyvector_ty, LLVMMatchType<1> ],
+                          [ IntrNoMem ]>;
```
Good spot, thanks. Amended now.

Lukacma commented on clang/include/clang/Basic/arm_sve.td:
```diff
@@ -1190,11 +1190,23 @@ def SVSUDOT_LANE_S : SInst<"svsudot_lane[_s32]", "ddqbi",  "i", MergeNone, "aarc
 
 let SVETargetGuard = "f32mm", SMETargetGuard = InvalidMode in {
-def SVMLLA_F32 : SInst<"svmmla[_f32]", "dddd","f", MergeNone, "aarch64_sve_fmmla">;
+def SVMLLA_F32 : SInst<"svmmla[_f32]", "dddd","f", MergeNone, "aarch64_sve_fmmla", [IsOverloadCvt]>;
```
I am bit confused about the inclusion of the `IsOverloadCvt` flag ? Is it necessary to add it here ?

Amichaxx committed:
Changed source type to fp8
Renamed IsOverloadCvt flag to IsOverloadFirstandLast for clarity

Amichaxx committed:
clang format

efriedma-quic commented:
See https://reviews.llvm.org/D113439 for similar discussion we had for RISC-V intrinsics.  Basically the same issues apply here.  The end result was that the intrinsics were reworked into their current form: the intrinsics take the mode as an argument, and an MIR pass converts the mode operand into explicit instructions to set the mode.

Lukacma commented on llvm/include/llvm/IR/IntrinsicsAArch64.td:
```suggestion
                          [LLVMMatchType<0>, llvm_nxv16i8_ty, llvm_nxv16i8_ty],
```

Amichaxx replied:
hmm I did try this and couldn't get it to work, as you can see:
`Intrinsic name not mangled correctly for type arguments! Should be: llvm.aarch64.sve.fp8.fmmla.nxv4f32`

Lukacma replied:
Yes that is because you also need to adjust clang intrinsic, as this intrinsic is overloaded on 1 type only now.

Amichaxx replied:
Ah apologies, thank you

Amichaxx committed:
- fp8_fmmla changes

Lukacma reviewed (approved):
LGTM. Just wait with merging this patch until ACLE is merged in case there are further changes there.
